application/controllers/template_controller.py

```diff
@@ -23,6 +24,17 @@ def get_template_by_id(template_id):
     return template
 
 
+def get_templates_by_classifiers(classifiers):
+    try:
+        templates = db.session.query(CommunicationTemplate).filter(CommunicationTemplate.classification == classifiers)\
+            .all()
+    except SQLAlchemyError:
+        logger.exception("Unable to retrieve template with classifiers", classifiers=classifiers)
+        raise DatabaseError("Unable to retrieve template with classifiers:  {}".format(json.dumps(classifiers)),
```

Comment:
I'm having a problem with Tox going insane about this even thought its 3.6

Reply:
SyntaxError or something else?

Reply:
flake8 problems
Last login: Mon Dec 11 08:43:47 on ttys001
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='2693245307'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py ...                                                                             [ 28%]
tests/views/test_classification_type_view.py .....F                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 48      3    94%   20-22
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             13      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         209     20    90%
Coverage HTML written to dir htmlcov


============================================================= FAILURES ==============================================================
________________________________ TestClassificationTypeView.test_upload_existing_classification_type ________________________________

self = <test_classification_type_view.TestClassificationTypeView testMethod=test_upload_existing_classification_type>

    def test_upload_existing_classification_type(self):
        # Given a ClassificationType exists in the database
        classification_type = "LEGAL_BASIS"
        self.__upload_classification_type(classification_type)
    
        # When the same classification type is uploaded
        response = self.client.post('/classificationtype/{}'.format(classification_type))
    
        # Then the service returns a 400 response
>       self.assertStatus(response, 400)

tests/views/test_classification_type_view.py:27: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py36/lib/python3.6/site-packages/flask_testing/utils.py:317: in assertStatus
    self.assertEqual(response.status_code, status_code, message)
E   AssertionError: 409 != 400 : HTTP Status 400 expected but got 409
------------------------------------------------------- Captured stdout call --------------------------------------------------------
2017-12-11 08:53.33 Uploading classification type: LEGAL_BASIS
2017-12-11 08:53.33 Uploaded new classification type classification_type=LEGAL_BASIS
2017-12-11 08:53.33 Uploading classification type: LEGAL_BASIS
2017-12-11 08:53.33 Attempted to upload an already existing classification type classification_type=LEGAL_BASIS
================================================ 1 failed, 20 passed in 1.46 seconds ================================================
ERROR: InvocationError: '/Users/waltoj/development/rm/rm-comms-template/.tox/py36/bin/pipenv run py.test --cov=application --cov-report html --cov-report term-missing'
______________________________________________________________ summary ______________________________________________________________
ERROR:   py36: commands failed
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='1842250570'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py ...                                                                             [ 28%]
tests/views/test_classification_type_view.py ......                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 48      3    94%   20-22
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             13      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         209     20    90%
Coverage HTML written to dir htmlcov


===================================================== 21 passed in 1.30 seconds =====================================================
______________________________________________________________ summary ______________________________________________________________
  py36: commands succeeded
  congratulations :)
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='4287093787'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py .FF                                                                             [ 28%]
tests/views/test_classification_type_view.py ......                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ....FF                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 65     16    75%   20-22, 49-65, 75-76
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             14      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         227     33    85%
Coverage HTML written to dir htmlcov


============================================================= FAILURES ==============================================================
___________________________ TestTemplateController.test_update_non_existent_template_creates_new_template ___________________________

self = <tests.controllers.test_template_controller.TestTemplateController testMethod=test_update_non_existent_template_creates_new_template>

    def test_update_non_existent_template_creates_new_template(self):
        # given the template doesn't exist in the database
        template_id = "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91"
        template_object = dict(id="cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91", label="test data", type="EMAIL",
                               uri="test-uri.com", classification={"GEOGRAPHY": "NI"})
    
        # When the template is updated
        is_created = template_controller.update_comms_template(template_id, template=template_object)
    
        # We receive an upload successful message with a boolean is_created
>       self.assertEquals(is_created, True)
E       AssertionError: False != True

tests/controllers/test_template_controller.py:30: AssertionError
------------------------------------------------------- Captured stdout call --------------------------------------------------------
2017-12-11 09:08.56 Updating template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
2017-12-11 09:08.56 Uploaded template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
______________________ TestTemplateController.test_upload_existing_template_raises_invalid_template_exception _______________________

self = <tests.controllers.test_template_controller.TestTemplateController testMethod=test_upload_existing_template_raises_invalid_template_exception>

    def test_upload_existing_template_raises_invalid_template_exception(self):
        template_id = "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91"
        template_object = dict(id="cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91", label="test data", type="EMAIL",
                               uri="test-uri.com", classification={"GEOGRAPHY": "NI"})
    
        # Given the object already exists in the database
        template_controller.update_comms_template(template_id, template=template_object)
    
        # When a template with the same id is uploaded
        with self.assertRaises(InvalidTemplateException):
>           template_controller.update_comms_template(template_id, template=template_object)
E           AssertionError: InvalidTemplateException not raised

tests/controllers/test_template_controller.py:18: AssertionError
------------------------------------------------------- Captured stdout call --------------------------------------------------------
2017-12-11 09:08.56 Updating template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
2017-12-11 09:08.56 Uploaded template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
2017-12-11 09:08.56 Updating template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
2017-12-11 09:08.56 Uploaded template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
_______________________________________________ TestTemplateView.test_update_template _______________________________________________

self = <test_template_view.TestTemplateView testMethod=test_update_template>

    def test_update_template(self):
        # Given there is a template in the database
        template_id = "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef89"
        data = dict(id=template_id, label="test data", type="EMAIL", uri="test-uri.com",
                    classification={"GEOGRAPHY": "NI"})
        self._upload_template(template_id, data)
    
        # When i update the template
        new_data = data.copy()
        new_data["label"] = "new label"
    
        response = self.client.put("/template/{}".format(template_id), content_type='application/json',
                                   data=json.dumps(new_data))
    
        # Then i receive a 200 response
>       self.assertStatus(response, 200)

tests/views/test_template_view.py:76: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py36/lib/python3.6/site-packages/flask_testing/utils.py:317: in assertStatus
    self.assertEqual(response.status_code, status_code, message)
E   AssertionError: 201 != 200 : HTTP Status 200 expected but got 201
------------------------------------------------------- Captured stdout call --------------------------------------------------------
2017-12-11 09:08.56 Creating template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef89
2017-12-11 09:08.56 Uploaded template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef89
2017-12-11 09:08.56 Updating template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef89
2017-12-11 09:08.56 Uploaded template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef89
______________________________ TestTemplateView.test_update_template_non_existent_template_creates_new ______________________________

self = <test_template_view.TestTemplateView testMethod=test_update_template_non_existent_template_creates_new>

    def test_update_template_non_existent_template_creates_new(self):
        # Given the template doesn't exist in the database
        template_id = "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef89"
        data = dict(id=template_id, label="test data", type="EMAIL", uri="test-uri.com",
                    classification={"GEOGRAPHY": "NI"})
    
        # When i update the template
        response = self.client.put("/template/{}".format(template_id), content_type='application/json',
                                   data=json.dumps(data))
    
        # Then i receive a 201 response
>       self.assertStatus(response, 201)

tests/views/test_template_view.py:89: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py36/lib/python3.6/site-packages/flask_testing/utils.py:317: in assertStatus
    self.assertEqual(response.status_code, status_code, message)
E   AssertionError: 200 != 201 : HTTP Status 201 expected but got 200
------------------------------------------------------- Captured stdout call --------------------------------------------------------
2017-12-11 09:08.56 Updating template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef89
2017-12-11 09:08.56 Uploaded template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef89
================================================ 4 failed, 17 passed in 1.47 seconds ================================================
ERROR: InvocationError: '/Users/waltoj/development/rm/rm-comms-template/.tox/py36/bin/pipenv run py.test --cov=application --cov-report html --cov-report term-missing'
______________________________________________________________ summary ______________________________________________________________
ERROR:   py36: commands failed
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='2528800745'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py ..F                                                                             [ 28%]
tests/views/test_classification_type_view.py ......                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 65     16    75%   20-22, 49-65, 75-76
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             14      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         227     33    85%
Coverage HTML written to dir htmlcov


============================================================= FAILURES ==============================================================
______________________ TestTemplateController.test_upload_existing_template_raises_invalid_template_exception _______________________

self = <tests.controllers.test_template_controller.TestTemplateController testMethod=test_upload_existing_template_raises_invalid_template_exception>

    def test_upload_existing_template_raises_invalid_template_exception(self):
        template_id = "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91"
        template_object = dict(id="cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91", label="test data", type="EMAIL",
                               uri="test-uri.com", classification={"GEOGRAPHY": "NI"})
    
        # Given the object already exists in the database
        template_controller.update_comms_template(template_id, template=template_object)
    
        # When a template with the same id is uploaded
        with self.assertRaises(InvalidTemplateException):
>           template_controller.update_comms_template(template_id, template=template_object)
E           AssertionError: InvalidTemplateException not raised

tests/controllers/test_template_controller.py:18: AssertionError
------------------------------------------------------- Captured stdout call --------------------------------------------------------
2017-12-11 09:11.51 Updating template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
2017-12-11 09:11.51 Uploaded template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
2017-12-11 09:11.51 Updating template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
2017-12-11 09:11.51 Uploaded template              id=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef91
================================================ 1 failed, 20 passed in 1.43 seconds ================================================
ERROR: InvocationError: '/Users/waltoj/development/rm/rm-comms-template/.tox/py36/bin/pipenv run py.test --cov=application --cov-report html --cov-report term-missing'
______________________________________________________________ summary ______________________________________________________________
ERROR:   py36: commands failed
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='96198046'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py ...                                                                             [ 28%]
tests/views/test_classification_type_view.py ......                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 65     14    78%   20-22, 49-65
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             14      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         227     31    86%
Coverage HTML written to dir htmlcov


===================================================== 21 passed in 1.29 seconds =====================================================
______________________________________________________________ summary ______________________________________________________________
  py36: commands succeeded
  congratulations :)
C02TP1NCG8WN:rm-comms-template waltoj$ git status
On branch get-by-classifier
Your branch is up to date with 'origin/get-by-classifier'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   application/controllers/classification_type_controller.py
	modified:   application/controllers/template_controller.py
	modified:   application/views/template_view.py
	modified:   tests/controllers/test_template_controller.py
	modified:   tests/views/test_classification_type_view.py
	modified:   tests/views/test_sessions.py

no changes added to commit (use "git add" and/or "git commit -a")
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='120893867'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py ...                                                                             [ 28%]
tests/views/test_classification_type_view.py ......                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 65     14    78%   20-22, 49-65
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             14      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         227     31    86%
Coverage HTML written to dir htmlcov


===================================================== 21 passed in 1.31 seconds =====================================================
______________________________________________________________ summary ______________________________________________________________
  py36: commands succeeded
  congratulations :)
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='1747974334'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py ...                                                                             [ 28%]
tests/views/test_classification_type_view.py ......                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 53      3    94%   20-22
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             14      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         215     20    91%
Coverage HTML written to dir htmlcov


===================================================== 21 passed in 1.29 seconds =====================================================
______________________________________________________________ summary ______________________________________________________________
  py36: commands succeeded
  congratulations :)
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='977946307'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py ...                                                                             [ 28%]
tests/views/test_classification_type_view.py ......                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 53      3    94%   20-22
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             14      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         215     20    91%
Coverage HTML written to dir htmlcov


===================================================== 21 passed in 1.29 seconds =====================================================
______________________________________________________________ summary ______________________________________________________________
  py36: commands succeeded
  congratulations :)
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='579389775'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 21 items                                                                                                                  

tests/controllers/test_classification_type_controller.py ...                                                                  [ 14%]
tests/controllers/test_template_controller.py ...                                                                             [ 28%]
tests/views/test_classification_type_view.py ......                                                                           [ 57%]
tests/views/test_info_view.py .                                                                                               [ 61%]
tests/views/test_sessions.py ..                                                                                               [ 71%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
application/__init__.py                                         1      0   100%
application/controllers/__init__.py                             0      0   100%
application/controllers/classification_type_controller.py      34      3    91%   14-16
application/controllers/info_controller.py                     12      2    83%   11-12
application/controllers/template_controller.py                 53      3    94%   20-22
application/error_handlers.py                                  22      7    68%   21-23, 35-38
application/models/__init__.py                                  0      0   100%
application/models/classification_type.py                       7      0   100%
application/models/models.py                                   28      0   100%
application/models/schema.py                                    2      0   100%
application/utils/__init__.py                                   0      0   100%
application/utils/database.py                                   2      0   100%
application/utils/exceptions.py                                11      0   100%
application/utils/logging.py                                    9      5    44%   14-29
application/views/__init__.py                                   0      0   100%
application/views/classification_type_view.py                  14      0   100%
application/views/info_view.py                                  6      0   100%
application/views/template_view.py                             14      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         215     20    91%
Coverage HTML written to dir htmlcov


===================================================== 21 passed in 1.31 seconds =====================================================
______________________________________________________________ summary ______________________________________________________________
  py36: commands succeeded
  congratulations :)
C02TP1NCG8WN:rm-comms-template waltoj$ git status
On branch get-by-classifier
Your branch is up to date with 'origin/get-by-classifier'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   application/controllers/classification_type_controller.py
	modified:   application/controllers/template_controller.py
	modified:   application/views/classification_type_view.py
	modified:   application/views/template_view.py
	modified:   tests/controllers/test_classification_type_controller.py
	modified:   tests/controllers/test_template_controller.py
	modified:   tests/views/test_classification_type_view.py
	modified:   tests/views/test_sessions.py
	modified:   tests/views/test_template_view.py

no changes added to commit (use "git add" and/or "git commit -a")
C02TP1NCG8WN:rm-comms-template waltoj$ git add application/
C02TP1NCG8WN:rm-comms-template waltoj$ git add tests/
C02TP1NCG8WN:rm-comms-template waltoj$ git commit -m "Splits out Put and Post for template controller into separate methods'
> "
[get-by-classifier f50ac58] Splits out Put and Post for template controller into separate methods'
 9 files changed, 64 insertions(+), 48 deletions(-)
C02TP1NCG8WN:rm-comms-template waltoj$ git push
Counting objects: 17, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (17/17), done.
Writing objects: 100% (17/17), 4.35 KiB | 4.35 MiB/s, done.
Total 17 (delta 10), reused 0 (delta 0)
remote: Resolving deltas: 100% (10/10), completed with 10 local objects.
To https://github.com/ONSdigital/rm-comms-template-service.git
   761863d..f50ac58  get-by-classifier -> get-by-classifier
C02TP1NCG8WN:rm-comms-template waltoj$ git branch
  delete-template
* get-by-classifier
  get-by-classifier-actual
  get-by-classifier2
  master
  uri-validation
C02TP1NCG8WN:rm-comms-template waltoj$ git checkout delete-template
Switched to branch 'delete-template'
Your branch is up to date with 'origin/delete-template'.
C02TP1NCG8WN:rm-comms-template waltoj$ git status
On branch delete-template
Your branch is up to date with 'origin/delete-template'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   application/views/template_view.py

no changes added to commit (use "git add" and/or "git commit -a")
C02TP1NCG8WN:rm-comms-template waltoj$ git diff
diff --git a/application/views/template_view.py b/application/views/template_view.py
index 3b6e65e..362b424 100644
--- a/application/views/template_view.py
+++ b/application/views/template_view.py
@@ -21,5 +21,5 @@ def get_template_by_id(template_id):
 @template_view.route('/template/<template_id>', methods=['DELETE'])
 def delete_template(template_id):
     is_deleted = template_controller.delete_comms_template(template_id)
-    http_code = 200 if is_deleted else 204
+    http_code = 200 if is_deleted else 404
     return Response(status=http_code)
C02TP1NCG8WN:rm-comms-template waltoj$ git add --all
C02TP1NCG8WN:rm-comms-template waltoj$ git status
On branch delete-template
Your branch is up to date with 'origin/delete-template'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   application/views/template_view.py

C02TP1NCG8WN:rm-comms-template waltoj$ git commit -m "Corrects HTTP code for deleting non-existent resource in template view'
> "
[delete-template 1965bc9] Corrects HTTP code for deleting non-existent resource in template view'
 1 file changed, 1 insertion(+), 1 deletion(-)
C02TP1NCG8WN:rm-comms-template waltoj$ git push
Counting objects: 5, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 465 bytes | 465.00 KiB/s, done.
Total 5 (delta 4), reused 0 (delta 0)
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To https://github.com/ONSdigital/rm-comms-template-service.git
   d91e69a..1965bc9  delete-template -> delete-template
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='1148340326'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
py36 runtests: commands[1] | pipenv run py.test --cov=application --cov-report html --cov-report term-missing
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
======================================================== test session starts ========================================================
platform darwin -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /Users/waltoj/development/rm/rm-comms-template, inifile:
plugins: cov-2.5.1
collected 12 items                                                                                                                  

tests/controllers/test_template_controller.py ...                                                                             [ 25%]
tests/views/test_info_view.py .                                                                                               [ 33%]
tests/views/test_sessions.py ..                                                                                               [ 50%]
tests/views/test_template_view.py ......                                                                                      [100%]

---------- coverage: platform darwin, python 3.6.3-final-0 -----------
Name                                             Stmts   Miss  Cover   Missing
------------------------------------------------------------------------------
application/__init__.py                              1      0   100%
application/controllers/__init__.py                  0      0   100%
application/controllers/info_controller.py          12      2    83%   11-12
application/controllers/template_controller.py      57      6    89%   20-22, 83-85
application/error_handlers.py                       18      7    61%   21-23, 28-31
application/models/__init__.py                       0      0   100%
application/models/models.py                        28      0   100%
application/models/schema.py                         1      0   100%
application/utils/__init__.py                        0      0   100%
application/utils/database.py                        2      0   100%
application/utils/exceptions.py                      9      0   100%
application/utils/logging.py                         9      5    44%   14-29
application/views/__init__.py                        0      0   100%
application/views/info_view.py                       6      0   100%
application/views/template_view.py                  15      0   100%
------------------------------------------------------------------------------
TOTAL                                              158     20    87%
Coverage HTML written to dir htmlcov


===================================================== 12 passed in 0.98 seconds =====================================================
______________________________________________________________ summary ______________________________________________________________
  py36: commands succeeded
  congratulations :)
C02TP1NCG8WN:rm-comms-template waltoj$ git add --all
C02TP1NCG8WN:rm-comms-template waltoj$ git commit -m 'test'
[delete-template eafa5c5] test
 1 file changed, 2 insertions(+), 2 deletions(-)
C02TP1NCG8WN:rm-comms-template waltoj$ git log
commit eafa5c5439c1b368b22cafc6a14fbe297293abf0 (HEAD -> delete-template)
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Mon Dec 11 09:42:12 2017 +0000

    test

commit 1965bc951f5916c495c3b0284555208070e3f722 (origin/delete-template)
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Mon Dec 11 09:35:49 2017 +0000

    Corrects HTTP code for deleting non-existent resource in template view'

commit d91e69a7de9ec8e565e1c022a6c93a457919d4d7
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Fri Dec 8 09:55:08 2017 +0000

    Adds delete template functionality

commit f2df10e35cd094686f9ee8b49e69fc2bf065e76a (origin/master, master)
Merge: 3c441d6 0505530
Author: jwgwalton <jwgwalton@gmail.com>
Date:   Fri Dec 8 11:55:56 2017 +0000

    Merge pull request #5 from ONSdigital/refactor
    
    Refactors code to remove unneeded class and tidy session creation

commit 05055303a456bfb8c9cedad32d847f59aa1beb17 (origin/refactor)
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Fri Dec 8 10:39:51 2017 +0000

    Key value pairs for logging
    
    rename function

commit 157e0287150bd4f374efa74c5f758da227d7706e
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Thu Dec 7 16:09:56 2017 +0000

    Refactors code to remove unneeded class and tidy session creation
    
    Removes unneeded body from 201 response on upload template

commit 3c441d68af4383766929bef856ff00f50700dd76
C02TP1NCG8WN:rm-comms-template waltoj$ git log
commit eafa5c5439c1b368b22cafc6a14fbe297293abf0 (HEAD -> delete-template)
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Mon Dec 11 09:42:12 2017 +0000

    test

commit 1965bc951f5916c495c3b0284555208070e3f722 (origin/delete-template)
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Mon Dec 11 09:35:49 2017 +0000

    Corrects HTTP code for deleting non-existent resource in template view'

commit d91e69a7de9ec8e565e1c022a6c93a457919d4d7
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Fri Dec 8 09:55:08 2017 +0000

    Adds delete template functionality

commit f2df10e35cd094686f9ee8b49e69fc2bf065e76a (origin/master, master)
Merge: 3c441d6 0505530
Author: jwgwalton <jwgwalton@gmail.com>
Date:   Fri Dec 8 11:55:56 2017 +0000

    Merge pull request #5 from ONSdigital/refactor
    
    Refactors code to remove unneeded class and tidy session creation

commit 05055303a456bfb8c9cedad32d847f59aa1beb17 (origin/refactor)
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Fri Dec 8 10:39:51 2017 +0000

    Key value pairs for logging
    
    rename function

commit 157e0287150bd4f374efa74c5f758da227d7706e
Author: Joseph Walton <jwgwalton@gmail.com>
Date:   Thu Dec 7 16:09:56 2017 +0000

    Refactors code to remove unneeded class and tidy session creation
    
    Removes unneeded body from 201 response on upload template

commit 3c441d68af4383766929bef856ff00f50700dd76
C02TP1NCG8WN:rm-comms-template waltoj$ git rebase -i f2df10e35cd094686f9ee8b49e69fc2bf065e76a
[detached HEAD f6d11bd] Adds delete template functionality
 Date: Fri Dec 8 09:55:08 2017 +0000
 4 files changed, 67 insertions(+), 3 deletions(-)
Successfully rebased and updated refs/heads/delete-template.
C02TP1NCG8WN:rm-comms-template waltoj$ git push -f
Counting objects: 12, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (12/12), done.
Writing objects: 100% (12/12), 2.36 KiB | 2.36 MiB/s, done.
Total 12 (delta 9), reused 0 (delta 0)
remote: Resolving deltas: 100% (9/9), completed with 9 local objects.
To https://github.com/ONSdigital/rm-comms-template-service.git
 + 1965bc9...f6d11bd delete-template -> delete-template (forced update)
C02TP1NCG8WN:rm-comms-template waltoj$ git branch
* delete-template
  get-by-classifier
  get-by-classifier-actual
  get-by-classifier2
  master
  uri-validation
C02TP1NCG8WN:rm-comms-template waltoj$ git status
On branch delete-template
Your branch is up to date with 'origin/delete-template'.

nothing to commit, working tree clean
C02TP1NCG8WN:rm-comms-template waltoj$ git checkout get-by-classifier2
Switched to branch 'get-by-classifier2'
Your branch is up to date with 'origin/get-by-classifier2'.
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='3083680627'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
./tests/views/test_template_view.py:104:21: E128 continuation line under-indented for visual indent
ERROR: InvocationError: '/Users/waltoj/development/rm/rm-comms-template/.tox/py36/bin/pipenv run flake8 .'
______________________________________________________________ summary ______________________________________________________________
ERROR:   py36: commands failed
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='3326781050'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
./tests/views/test_template_view.py:104:21: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:1:1: E902 TokenError: EOF in multi-line statement
./application/controllers/template_controller.py:33:104: E999 SyntaxError: invalid syntax
./application/controllers/template_controller.py:34:20: E225 missing whitespace around operator
./application/controllers/template_controller.py:35:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:38:1: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:39:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:41:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:43:70: E225 missing whitespace around operator
./application/controllers/template_controller.py:46:1: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:48:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:48:55: E225 missing whitespace around operator
./application/controllers/template_controller.py:50:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:52:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:52:22: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:52:24: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:54:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:55:72: E225 missing whitespace around operator
./application/controllers/template_controller.py:56:73: E225 missing whitespace around operator
./application/controllers/template_controller.py:58:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:58:10: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:58:12: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:59:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:59:9: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:59:11: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:60:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:60:8: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:60:10: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:61:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:61:19: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:61:21: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:62:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:62:11: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:62:13: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:64:13: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:15: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:40: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:59: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:71: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:81: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:101: E225 missing whitespace around operator
ERROR: InvocationError: '/Users/waltoj/development/rm/rm-comms-template/.tox/py36/bin/pipenv run flake8 .'
______________________________________________________________ summary ______________________________________________________________
ERROR:   py36: commands failed
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='3143712774'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
./application/controllers/template_controller.py:1:1: E902 TokenError: EOF in multi-line statement
./application/controllers/template_controller.py:33:104: E999 SyntaxError: invalid syntax
./application/controllers/template_controller.py:34:20: E225 missing whitespace around operator
./application/controllers/template_controller.py:35:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:38:1: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:39:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:41:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:43:70: E225 missing whitespace around operator
./application/controllers/template_controller.py:46:1: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:48:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:48:55: E225 missing whitespace around operator
./application/controllers/template_controller.py:50:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:52:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:52:22: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:52:24: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:54:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:55:72: E225 missing whitespace around operator
./application/controllers/template_controller.py:56:73: E225 missing whitespace around operator
./application/controllers/template_controller.py:58:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:58:10: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:58:12: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:59:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:59:9: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:59:11: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:60:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:60:8: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:60:10: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:61:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:61:19: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:61:21: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:62:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:62:11: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:62:13: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:64:13: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:15: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:40: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:59: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:71: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:81: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:101: E225 missing whitespace around operator
ERROR: InvocationError: '/Users/waltoj/development/rm/rm-comms-template/.tox/py36/bin/pipenv run flake8 .'
______________________________________________________________ summary ______________________________________________________________
ERROR:   py36: commands failed
C02TP1NCG8WN:rm-comms-template waltoj$ tox
GLOB sdist-make: /Users/waltoj/development/rm/rm-comms-template/setup.py
py36 inst-nodeps: /Users/waltoj/development/rm/rm-comms-template/.tox/dist/RM Comms Template-0.0.1.zip
py36 installed: attrs==17.3.0,certifi==2017.11.5,chardet==3.0.4,click==6.7,coverage==4.4.2,docker==2.6.1,docker-pycreds==0.2.1,flake8==3.5.0,Flask==0.12.2,Flask-Cors==3.0.3,Flask-SQLAlchemy==2.3.0,Flask-Testing==0.6.2,idna==2.6,itsdangerous==0.24,Jinja2==2.10,jsonschema==2.6.0,MarkupSafe==1.0,mccabe==0.6.1,nose==1.3.7,pew==1.1.1,pipenv==8.3.2,pluggy==0.6.0,psycopg2==2.7.1,py==1.5.2,pycodestyle==2.3.1,pyflakes==1.6.0,pytest==3.3.0,pytest-cov==2.5.1,requests==2.18.4,RM-Comms-Template==0.0.1,six==1.11.0,SQLAlchemy==1.1.10,SQLAlchemy-Utils==0.32.14,structlog==17.2.0,tox==2.9.1,tox-docker==1.0.0,urllib3==1.22,virtualenv==15.1.0,virtualenv-clone==0.2.6,websocket-client==0.44.0,Werkzeug==0.12.2
py36 runtests: PYTHONHASHSEED='2600043412'
py36 runtests: commands[0] | pipenv run flake8 .
Courtesy Notice: Pipenv found itself running within a virtual environment, so it will automatically use that environment, instead of creating its own for any project.
./application/controllers/template_controller.py:1:1: E902 TokenError: EOF in multi-line statement
./application/controllers/template_controller.py:34:97: E999 SyntaxError: invalid syntax
./application/controllers/template_controller.py:34:110: E225 missing whitespace around operator
./application/controllers/template_controller.py:35:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:38:1: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:39:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:41:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:43:70: E225 missing whitespace around operator
./application/controllers/template_controller.py:46:1: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:48:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:48:55: E225 missing whitespace around operator
./application/controllers/template_controller.py:50:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:52:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:52:22: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:52:24: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:54:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:55:72: E225 missing whitespace around operator
./application/controllers/template_controller.py:56:73: E225 missing whitespace around operator
./application/controllers/template_controller.py:58:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:58:10: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:58:12: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:59:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:59:9: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:59:11: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:60:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:60:8: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:60:10: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:61:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:61:19: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:61:21: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:62:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:62:11: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:62:13: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:64:13: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:15: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:40: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:59: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:71: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:81: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:101: E225 missing whitespace around operator

Reply:
flake 8 problems

./application/controllers/template_controller.py:1:1: E902 TokenError: EOF in multi-line statement
./application/controllers/template_controller.py:34:97: E999 SyntaxError: invalid syntax
./application/controllers/template_controller.py:34:110: E225 missing whitespace around operator
./application/controllers/template_controller.py:35:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:38:1: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:39:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:41:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:43:70: E225 missing whitespace around operator
./application/controllers/template_controller.py:46:1: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:48:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:48:55: E225 missing whitespace around operator
./application/controllers/template_controller.py:50:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:52:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:52:22: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:52:24: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:54:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:55:72: E225 missing whitespace around operator
./application/controllers/template_controller.py:56:73: E225 missing whitespace around operator
./application/controllers/template_controller.py:58:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:58:10: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:58:12: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:59:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:59:9: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:59:11: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:60:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:60:8: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:60:10: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:61:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:61:19: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:61:21: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:62:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:62:11: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:62:13: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:5: E128 continuation line under-indented for visual indent
./application/controllers/template_controller.py:64:13: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:15: E251 unexpected spaces around keyword / parameter equals
./application/controllers/template_controller.py:64:40: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:59: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:71: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:81: E225 missing whitespace around operator
./application/controllers/template_controller.py:64:101: E225 missing whitespace around operator
ERROR: InvocationError: '/Users/waltoj/development/rm/rm-comms-template/.tox/py36/bin/pipenv run flake8 .'

Reply:
@jwgwalton it almost looks like flake8 has been run for the first time...?